client/vaultclient/vaultclient.go

```diff
@@ -0,0 +1,933 @@
+	self, err := auth.LookupSelf()
+	if err != nil {
+		return fmt.Errorf("failed to lookup VaultClient's token: %v", err)
+	}
+
+	// Read and parse the fields
+	var data tokenData
+	if err := mapstructure.WeakDecode(self.Data, &data); err != nil {
+		return fmt.Errorf("failed to parse Vault token's data block: %v", err)
+	}
+
+	root := false
+	for _, p := range data.Policies {
+		if p == "root" {
+			root = true
+			break
+		}
+	}
+
+	if !data.Renewable && !root {
+		return fmt.Errorf("vault token is not renewable or root")
+	}
+
+	if data.CreationTTL == 0 && !root {
+		return fmt.Errorf("invalid lease duration of zero")
+	}
+
+	if data.TTL == 0 && !root {
+		return fmt.Errorf("token TTL is zero")
+	}
+
+	if !root && data.Role == "" {
+		return fmt.Errorf("token role name must be set when not using a root token")
+	}
+
+	data.Root = root
+	c.token = &data
+	return nil
+}
+
+// Stops the renewal loop of vault client
+func (c *vaultClient) Stop() {
+	if !c.config.Enabled || !c.running {
+		return
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.running = false
+	close(c.stopCh)
+}
+
+// DeriveToken takes in an allocation and a set of tasks and for each of the
+// task, it derives a vault token from nomad server and unwraps it using vault.
+// The return value is a map containing all the unwrapped tokens indexed by the
+// task name.
+func (c *vaultClient) DeriveToken(alloc *structs.Allocation, taskNames []string) (map[string]string, error) {
+	var result *multierror.Error
+
+	if !c.running {
+		result = multierror.Append(fmt.Errorf("vault client is not running"))
+		return nil, result
+	}
+
+	if alloc == nil {
+		result = multierror.Append(fmt.Errorf("nil allocation"))
+		return nil, result
+	}
+	if taskNames == nil || len(taskNames) == 0 {
+		result = multierror.Append(fmt.Errorf("missing task names"))
+		return nil, result
+	}
+
+	found := false
+	verifiedTasks := []string{}
+	// Check if the given task names actually exist in the allocation
+	for _, taskName := range taskNames {
+		found = false
+		for _, group := range alloc.Job.TaskGroups {
+			for _, task := range group.Tasks {
+				if task.Name == taskName {
+					found = true
+				}
+			}
+		}
+		if found {
+			verifiedTasks = append(verifiedTasks, taskName)
+		} else {
+			// Append the error for an invalid task name, but don't
+			// break out of the loop. Continue to process other
+			// tasks.
+			result = multierror.Append(result, fmt.Errorf("task %s not found in the allocation", taskName))
+		}
+	}
+
+	// DeriveVaultToken of nomad server can take in a set of tasks and
+	// creates tokens for all the tasks.
+	req := &structs.DeriveVaultTokenRequest{
+		NodeID:   c.node.ID,
+		SecretID: c.node.SecretID,
+		AllocID:  alloc.ID,
+		Tasks:    verifiedTasks,
+		QueryOptions: structs.QueryOptions{
+			Region:     c.region,
+			AllowStale: true,
+		},
+	}
+
+	// Derive the tokens
+	var resp structs.DeriveVaultTokenResponse
+	if err := c.RPC("Node.DeriveVaultToken", &req, &resp); err != nil {
```

Comment:
Lets pass in a DeriveVaultToken function to the vault client and then we can remove all the conn pool and rpc proxy stuff and it makes it easier to write tests